core/src/sigverify_shreds.rs

```diff
@@ -0,0 +1,395 @@
+                            let slot_end = slot_start + size_of::<u64>();
+                            let slot: Option<u64> =
+                                deserialize(&packet.data[slot_start..slot_end]).ok();
+                            if slot.is_some() && slot_leaders.get(&slot.unwrap()).is_some() {
+                                slot.unwrap()
+                            } else {
+                                std::u64::MAX
+                            }
+                        })
+                        .collect()
+                })
+                .collect()
+        })
+    });
+    let mut keys: HashMap<Pubkey, Vec<u64>> = HashMap::new();
+    for batch in slots.iter() {
+        for slot in batch.iter() {
+            let key = slot_leaders.get(slot).unwrap();
+            keys.entry(*key).or_insert(vec![]).push(*slot);
+        }
+    }
+    let mut pubkeys: PinnedVec<Pubkey> = recycler_pubkeys.allocate("shred_gpu_pubkeys");
+    let mut slot_to_key_ix = HashMap::new();
+    for (i, (k, slots)) in keys.iter().enumerate() {
+        pubkeys.push(*k);
+        for s in slots {
+            slot_to_key_ix.insert(s, i);
+        }
+    }
+    let mut offsets = recycler_offsets.allocate("shred_offsets");
+    slots.iter().for_each(|packet_slots| {
+        packet_slots.iter().for_each(|slot| {
+            offsets.push((slot_to_key_ix.get(slot).unwrap() * size_of::<Pubkey>()) as u32);
+        });
+    });
+    //HACK: Pubkeys vector is passed along as a `Packets` buffer to the GPU
+    //TODO: GPU needs a more opaque interface, which can handle variable sized structures for data
+    //Pad the Pubkeys buffer such that it is bigger than a buffer of Packet sized elems
+    let num_in_packets =
+        (pubkeys.len() * size_of::<Pubkey>() + (size_of::<Packet>() - 1)) / size_of::<Packet>();
+    trace!("num_in_packets {}", num_in_packets);
+    //number of bytes missing
+    let missing = num_in_packets * size_of::<Packet>() - pubkeys.len() * size_of::<Pubkey>();
+    trace!("missing {}", missing);
+    //extra Pubkeys needed to fill the buffer
+    let extra = (missing + size_of::<Pubkey>() - 1) / size_of::<Pubkey>();
+    trace!("extra {}", extra);
+    trace!("pubkeys {}", pubkeys.len());
+    for _ in 0..extra {
+        pubkeys.push(Pubkey::default());
+        trace!("pubkeys {}", pubkeys.len());
+    }
+    trace!("pubkeys {:?}", pubkeys);
+    trace!("offsets {:?}", offsets);
+    (pubkeys, offsets)
+}
+
+fn shred_gpu_offsets(
+    mut pubkeys_end: u32,
+    batches: &[Packets],
+    recycler_offsets: &Recycler<TxOffset>,
+) -> (TxOffset, TxOffset, TxOffset, Vec<Vec<u32>>) {
+    let mut signature_offsets = recycler_offsets.allocate("shred_signatures");
+    let mut msg_start_offsets = recycler_offsets.allocate("shred_msg_starts");
+    let mut msg_sizes = recycler_offsets.allocate("shred_msg_sizes");
+    let mut v_sig_lens = vec![];
+    for batch in batches {
+        let mut sig_lens = Vec::new();
```

Comment:
[optimization idea]: I don't know how it is likely the number of packets varies for each buckets. But at least we could preallocate (and possibly reuse) the most likely sized (probably for the case of batch with maximum number of packets) sig_lens in some way, assuming creating `Vec<i32>` is relatively costly in this context

Reply:
we have a recycler pattern already.  u32 vectors are pretty cheap, so its not going to thrash the heap that much.

Reply:
I see! Thanks for explaining!